### PR TITLE
LIT-3956 - wrapped-keys - Allow batch of keys to be stored atomically (storeEncryptedKeyBatch)

### PR DESCRIPTION
**Pull request by MaximusHaximus**
# Description
### New client API method `storeEncryptedKeyBatch()`, and associated param and result types
- Added new method to allow consumers of the API to persist a batch of keys to the backend storage service in a single, atomic, all-or-nothing request
- Updated existing API method `batchGeneratePrivateKeys` to use the new atomic method service-client method (`storePrivateKeyBatch()`) to write keys instead of calling multiple, independent `storePrivateKey` API calls in parallel.

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Ran localtests against `datil-test` (thus the TestNetworks) instance of our wrapped-keys REST API endpoints
- [x] Added new test for writing a batch of keys to the service with an _invalid_ entry in the array and verified that it throws the appropriate error from the backend

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts [hunk 1]
```diff
@@ -41,34 +41,40 @@ export async function batchGeneratePrivateKeys(
...
 
-  const results = await Promise.all(
-    actionResults.map(
-      async (result): Promise<BatchGeneratePrivateKeysActionResult> => {
-        const { generateEncryptedPrivateKey, network } = result;
+  const keyParamsBatch = actionResults.map((keyData) => {
+    const { generateEncryptedPrivateKey, network } = keyData;
+    return {
+      ...generateEncryptedPrivateKey,
+      keyType: getKeyTypeFromNetwork(network),
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:48]: This is for Solana/EVM?

File: packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts [hunk 2]
```diff
@@ -41,34 +41,40 @@ export async function batchGeneratePrivateKeys(
...
+    litNetwork: litNodeClient.config.litNetwork,
+  });
 
-        const signature = result.signMessage?.signature;
+  const results = actionResults.map(
+    (actionResult, ndx): BatchGeneratePrivateKeysActionResult => {
+      const {
+        generateEncryptedPrivateKey: { memo, publicKey },
+      } = actionResult;
+      const id = ids[ndx]; // Result of writes is in same order as provided
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:64]: What do the ids mean and can you remind me what is it used for? In the same order means EVM first then Solana, how is it determined or does it even matter a long as the process is atomic?

File: packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        dataToEncryptHash,
+        ciphertext,
+        keyType,
+      })
+    );
+
+  return storePrivateKeyBatch({
+    pkpAddress,
+    storedKeyMetadataBatch,
+    sessionSig: getFirstSessionSig(pkpSessionSigs),
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts:51]: Already extracted the sessionSig above
```suggestion
    sessionSig,
```

File: packages/wrapped-keys/src/lib/service-client/types.ts [hunk 4]
```diff
@@ -30,6 +30,14 @@ export interface StoreKeyParams extends BaseApiParams {
   >;
 }
 
+export interface StoreKeyBatchParams extends BaseApiParams {
+  pkpAddress: string;
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/service-client/types.ts:34]: Why is `pkpAddress` required as it isn't require in `StoreKeyParams`?

File: packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        publicKey,
+        memo,
+        dataToEncryptHash,
+        ciphertext,
+        keyType,
+      })
+    );
+
+  return storePrivateKeyBatch({
+    pkpAddress,
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts:49]: Why is this required isn't the `pkpAddress` extracted from the PKP sessionSig?

File: packages/wrapped-keys/src/lib/types.ts [hunk 6]
```diff
@@ -101,6 +101,31 @@ export interface StoreEncryptedKeyResult {
...
+
+/** Result of storing a batch of private keys in the wrapped keys backend service
+ * Includes an array of unique identifiers, which are necessary to get the encrypted ciphertext and dataToEncryptHash in the future
+ *
+ * @typedef StoreEncryptedKeyBatchResult
+ * @property { string } pkpAddress The LIT PKP Address that the key was linked to; this is derived from the provided pkpSessionSigs
+ * @property { string[] } ids Array of the unique identifiers (UUID V4) of the encrypted private keys in the same order provided
+ */
+export interface StoreEncryptedKeyBatchResult {
+  ids: string[];
```
Comment (DashKash54) [inline packages/wrapped-keys/src/lib/types.ts:125]: Are the ids used by the user in any way?

**Review by DashKash54 — COMMENTED** [review]
Looks good left a few comments

File: packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts (hunk 1 above)
Comment (MaximusHaximus) [reply packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:48]: Yep! it takes in the network string that is user-provided (evm or solana), and returns the type of key for that network (ed25519 or K256)

File: packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts (hunk 2 above)
Comment (MaximusHaximus) [reply packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:64]: The id is a UUIDv4 that we generate when storing into DynamoDB -- it has no direct meaning, it's just used as a unique identifier for that persisted key from then on.  When a consumer uses the `list` method on the wrapped keys client, it returns an array of all keys persisted for the session sig's PKP, including the ID.  When someone needs the encrypted data (dataToEncryptHash, cipherText) for a particular key, they the the ID of the key they want to the `getEncryptedKey()` method :+1:

File: packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts (hunk 5 above)
Comment (MaximusHaximus) [reply packages/wrapped-keys/src/lib/api/store-encrypted-key-batch.ts:49]: Good question! I checked into this, and it was a bit of a rabbit hole -- I realised that this method _doesn't_ need it -- but `list` and `get` do -- I _think_ it just ended up in all of the `service-client` methods dependencies parasitically even though we don't always need it.

This question unearthed an issue I hadn't noticed before, however -- we were importing the `getPkpAddressFromSessionSig()` method into `listPrivateKeyMetadata()` and `fetchPrivateKey()` _but we shouldn't have been_  the original design intention for `wrapped-keys` was that these operations would always be handled in the `api` layer and the result of them would be passed into the `service-client`, so we never ran those methods multiple times unnecessarily while servicing a `api` client call.

Thankfully, everything in the `service-client` is private so I fixed this problem and removed the `pkpAddress` parameter from the `service-client` parameter types that it was not needed for.

Fixing this encapsulation issue also exposed that we had imported the `service-client` and the internal util method into our `testImportPrivateKey` local-test -- so I fixed that too ;)

Good catch!

File: packages/wrapped-keys/src/lib/service-client/types.ts (hunk 4 above)
Comment (MaximusHaximus) [reply packages/wrapped-keys/src/lib/service-client/types.ts]: Good catch -- see https://github.com/LIT-Protocol/js-sdk/pull/686#discussion_r1797487662 - I resolved this

File: packages/wrapped-keys/src/lib/types.ts (hunk 6 above)
Comment (MaximusHaximus) [reply packages/wrapped-keys/src/lib/types.ts:125]: The _end-user_ generally would never see them -- they would see the `memo` field and probably public key -- the `id` is only used by developers when querying specific keys encrypted data :+1:

**Review by DashKash54 — APPROVED** [review]
LGTM!
